Source/WebCore/loader/cache/MemoryCache.cpp

```diff
@@ -157,23 +157,31 @@ void MemoryCache::revalidationSucceeded(CachedResource& revalidatingResource, co
 
     remove(revalidatingResource);
 
-    auto& resources = ensureSessionResourceMap(resource.sessionID());
-    auto key = std::make_pair(resource.url(), resource.cachePartition());
-
-    RELEASE_ASSERT(!resources.get(key));
-    resources.set(key, &resource);
-    resource.setInCache(true);
-    resource.updateResponseAfterRevalidation(response);
-    insertInLRUList(resource);
-    long long delta = resource.size();
-    if (resource.decodedSize() && resource.hasClients())
-        insertInLiveDecodedResourcesList(resource);
-    if (delta)
-        adjustSize(resource.hasClients(), delta);
+    auto& resources = ensureSessionResourceMap(resource->sessionID());
+    std::pair key { resource->url(), resource->cachePartition() };
+
+    auto addResult = resources.add(key, resource);
+    if (addResult.isNewEntry) {
+        resource->setInCache(true);
+        resource->updateResponseAfterRevalidation(response);
+        insertInLRUList(*resource);
+        long long delta = resource->size();
+        if (resource->decodedSize() && resource->hasClients())
+            insertInLiveDecodedResourcesList(*resource);
+        if (delta)
+            adjustSize(resource->hasClients(), delta);
+    } else {
+        // The resource was re-inserted in the cache during its revalidation.
+        // Ignore the revalidated resource and switch clients to the one that
+        // is already in the cache.
+        resource = addResult.iterator->value;
```

Comment:
Seems like we should assert that `addResult.iterator->value` is non-nullptr, then assign resource to the pointer referenced by the WeakPtr.

Reply:
It conflicted with my other MemoryCache patch. I am fixing now.